### DISCUSSION TITLE
How to solve the problem of garbled Chinese and English characters output from the terminal

### DISCUSSION THREAD

**Discussion by alex8788** [microsoft/vscode-discussions#1715] [Extension Development QnA] [answered]
The output from my terminal shows garbled text when printing both Chinese and English characters. I’ve tried various methods I found online (though the cases weren’t exactly the same as mine), but none of them worked. Below are some characteristics and test results I observed, and I hope someone can tell me how to solve this problem:

(C++)

- Directly using printf / cout to print: **No garbled text**
- Initializing a string and then printing: **Garbled text appears with printf, but not with cout**
- Declaring a string variable, reading input, and then printing: **Garbled text appears with scanf / printf, but not with cin / cout**
- The same string outputs different garbled text each time it's printed (which is very strange)

Based on the above characteristics, I believe the issue lies with scanf and printf.
![image](https://github.com/user-attachments/assets/deff5bea-6b25-4027-b19a-6da251a6d62b)
![image](https://github.com/user-attachments/assets/26317cce-73b8-4589-9f12-7238777d99ef)

Here are the methods/settings I've tried so far:

- Set VSCode's default encoding to UTF-8 and Big5
- Ensured that the program file is also encoded in UTF-8
- Ran the terminal command "chcp 65001"
- Followed the steps from [this article](https://blog.csdn.net/qq_59532828/article/details/141023391?spm=1001.2101.3001.6650.3&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7EYuanLiJiHua%7EPosition-3-141023391-blog-89823900.235%5Ev43%5Econtrol&utm_relevant_index=6)
- Modified command-line parameters in PowerShell and CMD settings (I don’t remember the exact tutorial I followed, but the changes included "chcp 65001")

=================================================

我的終端機輸出中文與英文時會變成亂碼，我試圖在網路上查詢各種方法(和我遇到的情況不太一樣)，但是都沒有用。以下是我觀察到的一些特徵與測試結果，希望有人能告訴我該如何解決:
(c++)
- 直接使用printf / cout打印: 不會出現亂碼
- 初始化一個字串再打印: 使用printf會出現亂碼而使用cout不會
- 宣告一個字串變數，讀取後再打印: 使用scanf / printf會出現亂碼而使用cin / cout不會
- 同樣的字串，每次打印出來的亂碼都不同(非常奇怪)
綜合上述幾個特徵，我認為問題出在scanf與printf
![code](https://github.com/user-attachments/assets/30e6e051-4dd9-478c-b4e2-aee08eb269e8)
![image](https://github.com/user-attachments/assets/ead9414e-6e2b-4994-9de8-bb2eb0ac1049)

目前我嘗試過的方法/設定:

- 將vscode預設編碼字元設定為utf-8,big5
- 確認程式檔案的編碼方式也是utf-8
- 在終端機執行命令"chcp 65001"
- 參考[該文章](https://blog.csdn.net/qq_59532828/article/details/141023391?spm=1001.2101.3001.6650.3&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7EYuanLiJiHua%7EPosition-3-141023391-blog-89823900.235%5Ev43%5Econtrol&utm_relevant_index=6)的做法
- 到power shell與cmd的設定頁面修改命令行參數(具體是哪個文章的教學我忘記了，不過替換的內容也包含"chcp 65001"

**Comment by robconery** [comment, marked as answer]
Hi @alex8788 - this discussion forum is primarily for extension development (see #5) and there might be a better public forum (like StackOverflow) to ask this question. That said - I also know that font choice is important for the terminal. I'm also wondering if there is a pointer issue here (I'm not a C++ person but when something weird happens with variables that's the first thing I think). Either way: it seems to be more of a C++ issue than a VS Code one.

**Comment by xiaoaiwhc** [comment]
Do you try these two options for g++ option?
-fexec-charset= your_character_set, for example: Big5 
-finput-charset= your_character_set
